### DISCUSSION TITLE
Publicodes dans un worker

### DISCUSSION THREAD

**Discussion by GroundZeroGreyHead** [publicodes/publicodes#94] [General]
lors de l'utilisation de publicode dans un web worker

```js
////worker js 
import { engine } from "./rules";
import { expose } from "threads/worker";

expose({
  workerComputeSituation(situation, evaluate) {
    if (typeof window !== "undefined") {
      return engine.setSituation(situation).evaluate(evaluate);
    }
  },
});
```

**Comment by mquandalle** [comment]
Il n'est pas encore possible de déplacer les calculs dans un worker, nous avions essayé de le faire dans betagouv/mon-entreprise#508 mais il y avait des problèmes avec l'interface du moteur qui n'était pas serialisable. Il reste encore un peu de travail mais ça devrait être possible bientôt. Il faudra aussi ajouter une section dans la documentation pour expliquer l'utilisation dans un worker.

Par curiosité, les calculs sont-ils trop lents pour être réalisés dans le process principal ?

**Comment by GroundZeroGreyHead** [comment]
C'est génial, donc les calculs continuent de bloquer l'interface utilisateur. c'est ainsi par exemple que nous effectuons le calcul.sur le call site je suis obligé de mettre des setTimeOut.
```js
import Engine from "publicodes";
import rules from "./rules";
import YAML from "yaml";
const RULES = YAML.stringify(rules);

export function computedSituation(options) {
  const value = engine
    .setSituation(options.situation)
    .evaluate(options.evaluate);
  // return formatValue(value, { displayedUnit: "" });
  return value.nodeValue;
```
il y a aussi des changements de configuration à apporter dans la configuration du webpack, rien de tout cela n'est précisé dans la documentation

**Comment by johangirod** [comment]
Je me permet d'ammender légèrement la réponse de @mquandalle. Si l'objet retourné par `evaluate` n'est pas serialisable en l'état, cette limitation ne s'applique pas à la chaine de charactère retournée par `formatValue`. 

Pour reprendre l'exemple : 
```js
////worker js 
import { engine } from "./rules";
import { expose } from "threads/worker";

expose({
  workerComputeSituation(situation, evaluate) {
    if (typeof window !== "undefined") {
      return formatValue(engine.setSituation(situation).evaluate(evaluate));
    }
  },
});
```
Cela devrait fonctionner.

> il y a aussi des changements de configuration à apporter dans la configuration du webpack, rien de tout cela n'est précisé dans la documentation

Très preneur de retours là dessus, il y a peut-être un ajustement à faire de notre côté lors de la compilation de la bibliothèque. Quel a été le changement à apporter ?

**Comment by GroundZeroGreyHead** [comment]
par exemple lors de l'utilisation de CRA il n'était pas possible de réutiliser les exemples dans[](https://mon-entreprise.fr/int%C3%A9gration/biblioth%C3%A8que-de-calcul)
sans surcharger la configuration du webpack pour les artefacts yaml certains de ce type
```js
const path = require("path");
const lodashCloneDeep = require("lodash/cloneDeep");
const ThreadsPlugin = require("threads-plugin");
const ruleChildren = (loader) =>
  loader.use ||
  loader.oneOf ||
  (Array.isArray(loader.loader) && loader.loader) ||
  [];

const findIndexAndRules = (rulesSource, ruleMatcher) => {
  let result = undefined;
  const rules = Array.isArray(rulesSource)
    ? rulesSource
    : ruleChildren(rulesSource);
  rules.some(
    (rule, index) =>
      (result = ruleMatcher(rule)
        ? {
            index,
            rules,
          }
        : findIndexAndRules(ruleChildren(rule), ruleMatcher))
  );
  return result;
};

const createLoaderMatcher = (loader) => (rule) =>
  rule.loader && rule.loader.indexOf(`${path.sep}${loader}${path.sep}`) !== -1;
const fileLoaderMatcher = createLoaderMatcher("file-loader");

const addBeforeRule = (rulesSource, ruleMatcher, value) => {
  const { index, rules } = findIndexAndRules(rulesSource, ruleMatcher);
  rules.splice(index, 0, value);
};

function rewireYAML(config) {
  const yamlLoader = {
    test: /\.ya?ml$/,
    type: "json", // Required by Webpack v4
    use: [{ loader: require.resolve("yaml-loader") }],
  };
  addBeforeRule(config.module.rules, fileLoaderMatcher, yamlLoader);

  return config;
}

module.exports = {
  webpack: function (config, env) {
    // config.module.rules.push({
    //   test: /\.worker\.js$/,
    //   use: { loader: "worker-loader" },
    // });
    const workerExtension = /\.worker\.js$/;
    function isBabelLoader(rule) {
      return rule.loader && rule.loader.indexOf("babel-loader") !== -1;
    }

    function findBabelLoader(rule) {
      if (isBabelLoader(rule)) {
        return rule;
      }

      if (Array.isArray(rule.use) && rule.use.find(isBabelLoader)) {
        return rule;
      }

      return Array.isArray(rule.oneOf) && rule.oneOf.find(isBabelLoader);
    }

    function searchRules(rules) {
      for (let i = 0; i < rules.length; i++) {
        const babelRule = findBabelLoader(rules[i]);
        if (babelRule) {
          return babelRule;
        }
      }

      return {};
    }

    const babelLoader = searchRules(config.module.rules);

    const workerLoader = lodashCloneDeep(babelLoader);

    workerLoader.test = workerExtension;
    workerLoader.use = [
      "worker-loader",
      {
        // Old babel-loader configuration goes here.
        loader: workerLoader.loader,
        options: workerLoader.options,
      },
    ];
    delete workerLoader.loader;
    delete workerLoader.options;

    babelLoader.exclude = (babelLoader.exclude || []).concat([workerExtension]);

    config.module.rules.push(workerLoader);

    config = rewireYAML(config, env);
    config.output.globalObject = `(() => {
          if (typeof self !== 'undefined') {
              return self;
          } else if (typeof window !== 'undefined') {
              return window;
          } else if (typeof global !== 'undefined') {
              return global;
          } else {
              return Function('return this')();
          }
      })()`;
    if (!config.plugins) {
      config.plugins = [];
    }
    config.plugins.push(new ThreadsPlugin());
    return config;
  },
};
```
Tu fais probablement la même chose
J'ai essayé de réutiliser votre implémentation typescript mais les types dotName n'étaient pas présents.

**Comment by GroundZeroGreyHead** [comment]
Avec une implémentation threaded, nous obtenons la stack d'erreurs suivante.
Pas de problème avec l'erreur de clonage de données en raison de serialiase puisque engine n'est pas passé en tant que paramètre de message.
```js
ServerStyleSheet.js:66 Uncaught ReferenceError: window is not defined
    at Module.<anonymous> (ServerStyleSheet.js:66)
    at Module../node_modules/styled-components/dist/styled-components.browser.esm.js (ServerStyleSheet.js:66)
    at __webpack_require__ (bootstrap:19)
    at webpackUniversalModuleDefinition (index.js:2)
    at Object../node_modules/publicodes/dist/index.js (index.js:10)
    at __webpack_require__ (bootstrap:19)
    at Module../node_modules/babel-loader/lib/index.js?!./node_modules/eslint-loader/dist/cjs.js?!./src/Worker.js (Worker.js:1)
    at __webpack_require__ (bootstrap:19)
    at bootstrap:83
    at bootstrap:83
```
Implémentation:
```js
worker.js
import formatValue from "publicodes";
import { engine } from "./rules";
import { expose } from "threads/worker";

expose({
  workerComputeSituation(situation, evaluate) {
    return formatValue(engine.setSituation(situation).evaluate(evaluate));
  },
});

main thread
  async pWorker(situation, evaluate) {
    const worker = await spawn(new Worker("./Worker"));
    const compute = await worker.workerComputeSituation(situation, evaluate);
    await Thread.terminate(worker);
  }
```

**Comment by johangirod** [comment]
> par exemple lors de l'utilisation de CRA il n'était pas possible de réutiliser les exemples dans
sans surcharger la configuration du webpack pour les artefacts yaml certains de ce type

Effectivement, si on veut pouvoir parser le yaml à l'import, il faut modifier la config webpack de CRA. Mais il est également possible de fournir au constructeur d'Engine les règles non parsée. C'est ce qui est fait dans les exemples d'[utilisation des règles de calcul](https://mon-entreprise.fr/int%C3%A9gration/biblioth%C3%A8que-de-calcul), avec une concaténation des [règles avant le parsing du yaml dans le moteur]( https://codesandbox.io/s/mon-entreprise-exemple-1-2dpct?from-embed=&file=/src/rules/index.js). 

Ce n'est pas forcément l'idéal, puisque les erreur syntaxique propre au yaml n'indiqueront pas le fichier concerné, mais cela évite d'avoir à modifier la conf webpack.

### Stack d'erreur

```
ServerStyleSheet.js:66 Uncaught ReferenceError: window is not defined
    at Module.<anonymous> (ServerStyleSheet.js:66)
    at Module../node_modules/styled-components/dist/styled-components.browser.esm.js (ServerStyleSheet.js:66)
    at __webpack_require__ (bootstrap:19)
    at webpackUniversalModuleDefinition (index.js:2)
    at Object../node_modules/publicodes/dist/index.js (index.js:10)
    at __webpack_require__ (bootstrap:19)
    at Module../node_modules/babel-loader/lib/index.js?!./node_modules/eslint-loader/dist/cjs.js?!./src/Worker.js (Worker.js:1)
    at __webpack_require__ (bootstrap:19)
    at bootstrap:83
    at bootstrap:83
```

Effectivement, cela vient de l'utilisation de l'objet window dans les styled-component. Ce problème sera réglé avec betagouv/mon-entreprise#1070, prévu pour cette fin d'année. En attendant, le seul hack existant est de mocker l'objet window pour que l'evaluation puisse avoir lieu.

Pour info, j'avais essayé de déplacer les calcul dans le worker pour mon-entreprise, avant d'abandonner suite aux limites rencontrée. Voici les conclusions que j'en avais tiré : betagouv/mon-entreprise#508 

EDIT : mauvais lien

**Comment by GroundZeroGreyHead** [comment]
> Effectivement, cela vient de l'utilisation de l'objet window dans les styled-component. Ce problème sera réglé avec betagouv/mon-entreprise#1070, prévu pour cette fin d'année. En attendant, le seul hack existant est de mocker l'objet window pour que l'evaluation puisse avoir lieu.

Comment proposez-vous que je fasse cela dans le contexte du worker thread ? 

> Pour info, j'avais essayé de déplacer les calcul dans le worker pour mon-entreprise, avant d'abandonner suite aux limites rencontrée. Voici les conclusions que j'en avais tiré : betagouv/mon-entreprise#1070

Comment pouvez-vous empêcher un processus de longue durée de ne pas bloquer votre main

**Comment by johangirod** [comment]
> Comment pouvez-vous empêcher un processus de longue durée de ne pas bloquer votre main

 En attendant, nous avons choisi d'ajouter un [debounce](https://github.com/betagouv/mon-entreprise/blob/master/mon-entreprise/source/utils.ts#L9-L18) aux entrées utilisateurs pour ne pas bloquer le main thread immédiatement. Ce n'est pas idéal, et nous en avons conscience, d'où nos tentatives pour régler le problème (#508 betagouv/mon-entreprise#1208).

Nous avons maintenant une bonne idée des étapes à accomplir, et nous espérons pouvoir régler ce problème d'ici à la fin d'année.

> Comment proposez-vous que je fasse cela dans le contexte du worker thread ?

J'imagine en commençant par définir un objet global window dans le fichier `worker.js` 
```js
self.window = {}
```
Mais c'est clairement un hack en attendant mieux.

**Comment by mquandalle** [comment]
Pour information la dernière version beta de Publicodes (`beta.7`) retire la dépendance à React qui rendait la librairie inutilisable dans un web worker. Je n'ai pas encore essayé d’exécuter Publicodes dans un worker donc il est possible qu'il existe encore des blocages, mais on va essayer de faire la modification sur mon-entreprise.fr très prochainement.

**Comment by mquandalle** [comment]
On doit pouvoir utiliser https://github.com/GoogleChromeLabs/comlink pour interfacer facilement l'API publicodes avec une instance dans un worker.
